### PR TITLE
Prefer end-tasks with low numbers of dependencies when ordering

### PR DESCRIPTION
**Pull request by TomAugspurger**
Starting at https://github.com/dask/dask-ml/issues/206#issuecomment-395869929

Just a test so far. Coming back to this later this afternoon.

### CODE REVIEW

File: dask/tests/test_order.py [hunk 1]
```diff
@@ -409,3 +409,95 @@ def test_string_ordering_dependents():
                  ('a', 1): 1,
                  ('a', 2): 2,
                  ('a', 3): 3}
+
+
+def test_prefer_short_narrow():
```
Comment (mrocklin) [inline dask/tests/test_order.py]: I recommend using the abcde fixture here to use that for key names.  It changes order to ensure that we're not entirely dependent on the fact that, `'a' < 'b'` for example.

**Comment by TomAugspurger** [comment]
Just leaving some notes to pick up later.

The most minimal change, which breaks other tests, is 

```diff
@@ -139,9 +150,13 @@ def order(dsk, dependencies=None):
         deps = [d for d in dependents[item]
                 if d not in result and not (d in seen and waiting[d])]
         if len(deps) < 1000:
-            deps = sorted(deps, key=dependents_key, reverse=True)
+            deps = sorted(deps, key=dependents_key, reverse=False)

```

with `reverse=True`, we put `('a', 1)` (which represents the last data block of `X`) next on the stack / result list.

This fails

- `test_gh_3505`
- `test_prefer_short_dependents`
- `test_nearest_neighbor`
- `test_string_ordering_dependents`

Investigating why later.

**Comment by TomAugspurger** [comment]
Briefly, though, it seems like that `reverse=True` is there so we pick tasks with shorter task chains

With `reverse=True` (master)

mydask


With `reverse=False`

mydask

**Comment by mrocklin** [comment]
The objective behind that choice is to allow c to be collected early on.  Otherwise it sticks around while we run the longer chain.

**Comment by jakirkham** [comment]
That makes sense. It seems where it gets thrown off is when we add a shared vertex. In these cases having a short chain doesn't matter as it will still be dependent on the longer chains to finish (suggesting the longest chain of a shared vertex should get highest priority). Is there a good way to include some logic regarding shared vertices and their dependents?

**Comment by TomAugspurger** [comment]
I think a rule of "adding a common *dependent*" shouldn't change the ordering is a good one. Still not sure how we would actually do that quickly though.

**Comment by mrocklin** [comment]
> Still not sure how we would actually do that quickly though.

Near-optimal task scheduling in linear time is hard :)

**Comment by TomAugspurger** [comment]
Picking this up again today.

I think the cleanest example is the following:

```python
a, b, c, d, e, = 'abcde'

def f(*args): pass

good = {
        (a, 0): 0,
        (b, 0): 0,
        (c, 0): 0,
        (c, 1): (f, (c, 0), (a, 0), (b, 0)),
        (a, 1): 1,
        (b, 1): 1,
        (c, 2): (f, (c, 1), (a, 1), (b, 1)),
}
 

bad = {
    ('ab'): 0,
    (a, 0): (f, 'ab', 0, 0),
    (b, 0): (f, 'ab', 0, 1),
    (c, 0): 0,
    (c, 1): (f, (c, 0), (a, 0), (b, 0)),
    (a, 1): (f, 'ab', 1, 0),
    (b, 1): (f, 'ab', 1, 1),
    (c, 2): (f, (c, 1), (a, 1), (b, 1)),
}
```

Diagrammatically, that's 

```
    Good:

                    c2
                   / \ \
                  /   \ \
                c1     \ \
              / | \     \ \
             c  a0 b0   a1 b1

    Bad:

                    c2
                   / \ \
                  /   \ \
                c1     \ \
              / | \     \ \
             c  a0 b0   a1 b1
                   \ \   / /
                    \ \ / /
                      a-b
```

And visually, it's

good:


mydask

bad:

mydask



cc @eriknw if this ordering issue interests you.

**Comment by mrocklin** [comment]
I took a look at this problem.  Stepping through order it seems like it does correctly identify that c1 is a higher priority than a1, b2, then it goes to a0 or b1, and from there to a-b.  Everything here so far is correct. 

What happens next though is unfortunate, It looks up from ab and sees both a0/b0 and a1/b1 and it decides that it would prefer what it perceives to be the path that will finish the soonest.  Its proxy for "soonest" is the number of total dependents, so it chooses a1/b1, which in this case is wrong.

Number of total dependents is a poor (but cheap!) proxy for soonest-finishing-path.  It's wrong because all paths are ending in the same place here, c2.  In this case we would actually prefer to take the branch that has the most work between itself and the end point.  I wonder if something like the following would be productive (I give this a 30% chance of being a good idea):

1.  For each node compute the maximum total_dependencies among all of its dependents.  In this case all nodes would have the same value around 10.  I'm calling this `max_dependencies` for now.  Computing this would be similar to the current `ndependents/ndependencies` functions, except that they would probably also consume the `total_dependents/total_dependencies` as inputs.
2.  In the `dependents_key` function use something like `max_dependents[key] - total_dependents[key]` to select for nodes that are close to finished.

Anyway, just thinking out loud, it's something to try.

**Comment by mrocklin** [comment]
Also, FWIW, for debugging I like putting the following just after the `while stack:` line

```python
print(stack)
import pdb; pdb.set_trace()
```

I find that this is a good way to see how things progress, while being able to either easily step through the algorithm iteration by iteration (with `c`) or investigate what happens next as necessary.

**Comment by TomAugspurger** [comment]
Thanks.

My mind went to something like "number of dependents, not included the dependents of my sibling tasks", but it wasn't clear to be how "siblings" would be tracked in `ndependents`, or whether that even generalizes beyond this problem.

Your proposal sounds simpler. I can try it out.

> Anyway, just thinking out loud, it's something to try.

I appreciate it.

**Comment by TomAugspurger** [comment]
FYI, for fun I was playing around with changing `dask_ml._partial.fit` to use the distributed scheduler and user-defined priorities. That seemed to help, but it'd be beneficial to fix it here.

**Comment by mrocklin** [comment]
Yeah, I'd prefer that we focus efforts on dask.order if possible, even if
it's a harder problem to improve.

**Comment by mrocklin** [comment]
@TomAugspurger if you have a chance can I ask you to try your workloads with this branch?  If things work out then I'll modify docs and tune things a bit.

**Comment by TomAugspurger** [comment]
Seems to be ideal on the small example I showed on Friday.

out


I'll try it out on a larger problem.

**Comment by TomAugspurger** [comment]
As a side, note, it seems like the previous `fit`s are no longer held in memory. I don't think that change was in dask-ml.

**Comment by mrocklin** [comment]
Hrm, it does *generate* a lot of data though while its waiting for the
model fit to proceed.  This is a separate problem though.

**Comment by mrocklin** [comment]
OK, I've updated the module-level docstring

**Comment by mrocklin** [comment]
@TomAugspurger your review of the module-level docstring would be particularly welcome

File: dask/order.py [hunk 2]
```diff
@@ -16,45 +16,58 @@
...
-
-1.  Number of dependents (smaller is better)
-2.  Number of total downstream dependencies (larger is better)
-3.  Name of the task itself, as a tie breaker
+There are several ways in which we might order our keys.  This is a nuanced
+process that has to take into account many different kinds of workflows, and
+operate efficiently in linear time.  We strongly recommend that readers look at
+the docstrings of tests in dask/tests/test_order.py.  These tests usually have
+graph types laid out very carefully to show the kinds of situations that often
+arise.
```
Comment (TomAugspurger) [inline dask/order.py]: ", and the order we would *like* to be determined`."

File: dask/order.py [hunk 3]
```diff
@@ -16,45 +16,58 @@
...
+
+
+Policy
+------
+
+Work towards *small goals* with *big steps*.
+
+1.  **Small goals**: prefer tasks whose final dependents have few dependencies.
+
+    By final dependent we mean a task that depends on this task that is the end
```
Comment (TomAugspurger) [inline dask/order.py]: I don't quite follow this first sentence.

File: dask/order.py [hunk 4]
```diff
@@ -140,37 +155,55 @@ def dependents_key(x):
...
+    min_dependencies: Dict[key, int]
     """
     result = dict()
+    min_result = dict()
     num_needed = {k: len(v) for k, v in dependents.items()}
     current = {k for k, v in num_needed.items() if v == 0}
     while current:
         key = current.pop()
         result[key] = 1 + sum(result[parent] for parent in dependents[key])
+        try:
```
Comment (TomAugspurger) [inline dask/order.py]: Is the `try / except` to handle the empty case? If so, `min` takes a `default=` argument. So

```python
min_result[key] = min(min_result[parent] for parent in dependents[key], default=total_dependencies[key])
```

Comment (mrocklin) [reply dask/order.py]: While more verbose, try-except is often performance optimal when the try case is much more common.  

```python
In [1]: d = {'x': 1}

In [2]: %timeit min([], default=d['x'])
378 ns ± 16.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: %timeit min([1, 2], default=d['x'])
413 ns ± 1.26 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit min([1, 2])
166 ns ± 0.642 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [6]: %%timeit
   ...: try:
   ...:     min([1, 2])
   ...: except TypeError:
   ...:     d['x']
   ...:     
178 ns ± 2.74 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [8]: %%timeit
   ...: try:
   ...:     min([])
   ...: except ValueError:
   ...:     d['x']
   ...:     
493 ns ± 4.52 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

This difference is small, but we do actually care about things at the microsecond level here.

Comment (TomAugspurger) [reply dask/order.py]: Huh, I'm surprised by the overhead added by the default argument. Good to know.

**Comment by mrocklin** [comment]
it's probably the dictionary access d[x]

**Comment by TomAugspurger** [comment]
Playing with this larger example, 16GB, 200 blocks.

```python
from distributed import Client
client = Client(processes=False)
client

import os
from dask_ml.datasets import make_classification
from dask_ml.wrappers import Incremental
from sklearn.linear_model import SGDClassifier
import dask.array as da

X, y = make_classification(n_samples=2_000_000, n_features=1000, chunks=10_000, random_state=0)

if not os.path.exists("Xy.zarr"):
    Xy = da.concatenate([X, y.reshape(-1, 1)], axis=1)
    Xy.to_zarr('Xy.zarr', overwrite=True)

Xy = da.from_zarr("Xy.zarr")
X = Xy[:, :-1]
y = Xy[:, -1]

clf = SGDClassifier(random_state=0)
inc = Incremental(clf, scoring='accuracy')

%time inc.fit(X, y, classes=[0, 1])
```

Status pane: https://streamable.com/upsan

We get some nice overlap between data IO and model training. We're loading the next blocks while we train on previous ones.

However, I *think* the scheduler decides to process all the load zarrs as soon as possible. Because we haven't released enough yet, since we haven't gone through all the `fit`s, we end up spilling to disk. Is this the desired behavior?

**Comment by mrocklin** [comment]
Yeah, that's what I meant to highlight in https://github.com/dask/dask/pull/3588#issuecomment-400402803 .  It's a problem, but a separate problem.  I've been thinking about it a bit recently and haven't come to any very strong conclusions.

**Comment by mrocklin** [comment]
I plan to merge this early tomorrow if there are no further comments.

**Comment by jakirkham** [comment]
> However, I *think* the scheduler decides to process all the load zarrs as soon as possible. Because we haven't released enough yet, since we haven't gone through all the `fit`s, we end up spilling to disk. Is this the desired behavior?

Should we move this to a new issue or is there one already?

File: dask/tests/test_order.py [hunk 5]
```diff
@@ -232,16 +232,22 @@ def test_order_doesnt_fail_on_mixed_type_keys(abcde):
...
+    orig = x = da.random.normal(size=(A, B), chunks=(1, None))
     for _ in range(2):
         y = (x[:, None, :] * x[:, :, None]).cumsum(axis=0)
         x = x.cumsum(axis=0)
     w = (y * x[:, None]).sum(axis=(1,2))
 
     dsk = dict(w.__dask_graph__())
     o = order(dsk)
+    # dask.visualize(dsk, color='order', node_attr={'penwidth': '6'},
+    #                filename='dask.pdf')
```
Comment (jakirkham) [inline dask/tests/test_order.py]: Should this be dropped?

Comment (mrocklin) [reply dask/tests/test_order.py]: I've added these lines the last few times I've played with order.  I'm inclined to leave them in for myself or others.  The image made by this test is particularly valuable to check correctness.

**Comment by mrocklin** [comment]
I would like to merge this.  Any objections?

**Comment by mrocklin** [comment]
> Should we move this to a new issue or is there one already?

I did a quick look and didn't see anything.  I wouldn't be surprised if it has come up a few times already though in various issues.

It's the sort of thing that, unfortunately, has a bus factor of 1.  I've mostly been thinking about it on my own.

**Comment by mrocklin** [comment]
Merging in a few hours if there are no further comments

File: dask/tests/test_order.py (hunk 5 above)
Comment (jakirkham) [reply dask/tests/test_order.py]: Generally don't find commented code to be a good idea to keep around (especially in a nice library like Dask). It is seldom updated or used for that matter. Also it raises questions for later users as to its value (at which point it may end up being scrapped anyways). Not to mention it makes the code base a bit uglier. If the code is simple like this, it is easy for anyone to add should it be needed. Would recommend dropping it. However if you feel strongly about keeping it, would encourage including it as part of the test explicitly (uncommented) so it doesn't fall victim to the aforementioned issues. Given how easy this is to solve, would encourage picking one of these two options and addressing it before merging.

**Comment by jakirkham** [comment]
> > > However, I *think* the scheduler decides to process all the load zarrs as soon as possible. Because we haven't released enough yet, since we haven't gone through all the `fit`s, we end up spilling to disk. Is this the desired behavior?
> >
> > Should we move this to a new issue or is there one already?
>
> I did a quick look and didn't see anything.  I wouldn't be surprised if it has come up a few times already though in various issues.
>
> It's the sort of thing that, unfortunately, has a bus factor of 1.  I've mostly been thinking about it on my own.

That's fair. Have no strong feelings about this. So happy with whatever you decide here. Though an open issue might track relevance if nothing else. Who knows maybe there are users interested in working on this problem. 😉

File: dask/tests/test_order.py (hunk 5 above)
Comment (mrocklin) [reply dask/tests/test_order.py]: As you like.  Removed

**Comment by mrocklin** [comment]
Anything else to do here @jakirkham ?

**Comment by jakirkham** [comment]
Not from me. :) Feel free to merge it when ready.

**Comment by estebanag** [comment]
I've been looking into the effects of the changes made to [dask/order.py](https://github.com/dask/dask/blob/master/dask/order.py) from release **0.16.1** up to [this commit](https://github.com/dask/dask/commit/f0deae99d3ddd18bbfe6e51d521ce760bf303cb4).

In particular, I've noticed that from release **0.17.2** onwards the memory footprint can increase significantly in some cases (see this [issue](https://github.com/dask/dask/issues/3524)). My workaround is to take an older version of [dask/order.py](https://github.com/dask/dask/blob/0.17.1/dask/order.py) and replace it on my master branch.

I wonder whether the task prioritization strategy should be parametrized (after defining some common use cases) instead of trying to have a "one size fits all" strategy.

**Comment by mrocklin** [comment]
I would be concerned that maintaining several ordering algorithms within the codebase would increase maintenance burden.  I'm inclined to try to find one way that works well for everyone.  However, I'd be open to making this more pluggable so that others could experiment with their own.

**Comment by jakirkham** [comment]
> However, I _think_ the scheduler decides to process all the load zarrs as soon as possible. Because we haven't released enough yet, since we haven't gone through all the `fit`s, we end up spilling to disk. Is this the desired behavior?

Wound up running into some problems recently due to this behavior and ended up back in this PR. 😄

So have gone ahead and raised issue ( https://github.com/dask/dask/issues/4310 ) to track it. AFAICT I'm seeing the same issue though may have a *slightly* simpler example of it.
